exp/etcdrestore/webhooks/rbac.go

```diff
@@ -0,0 +1,60 @@
+/*
+Copyright © 2023 - 2024 SUSE LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package webhooks
+
+import (
+	"context"
+	"fmt"
+
+	authv1 "k8s.io/api/authorization/v1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+func validateRBAC(ctx context.Context, cl client.Client, clusterName, clusterNamespace string) error {
+	admissionRequest, err := admission.RequestFromContext(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get admission request from context: %w", err)
+	}
+
+	sar := authv1.SubjectAccessReview{
+		Spec: authv1.SubjectAccessReviewSpec{
+			ResourceAttributes: &authv1.ResourceAttributes{
```

Comment:
Doesn’t this need a check for access to `Machine` as well? Maybe there needs to be a user group a user needs to belong to and validated against.

Reply:
If the user is allowed to administer `Cluster` object, can we assume they have access to machines? `SubjectAccessReview` should handle all validation on API server side, we provide it all information we have like user name, group and resource we'd like to validate against.
I'd like to avoid doing multiple SARs if possible assuming that there might be a lot of ETCDMachineSnapshots.

Reply:
My take on this is to have a `etcd-snapshot-restore` cluster role, which a user has to be a part of to perform operation. This does not mean making more `SAR` requests.

Alternatively, what prevents here to check for `Machine` object in this request too?

Reply:
We started by thinking of including custom RBAC verbs like "etcd-snapshot" or "etcd-restore" and then checking for them in SAR. Then we decided it easier to restrict etcd snapshot/restore to admin users initially and it can be easily changed later if we get any requests from users.

Reply:
For more context, a verb in `Role/ClusterRole` can be any word for example:
```
  verbs: ["get", "list", "watch", "create", "update", "patch", "delete", "mycustomverb"]
```
It's up to us later, how to handle it in webhooks/controllers. For example, we might want more granular RBAC rules for some resources and allow only specific field change for some users in a resource

Reply:
It is a new pattern, I was not aware of. My suggestion was to stick to classic RBAC management roles, where a user may be `admin` or `view` or a member of a custom `backup`, `backuprestore` group. It is an alternative, which came to me since I was not a part of previous discussion.

Makes sense to me.